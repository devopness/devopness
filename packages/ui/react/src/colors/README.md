# src/colors

Devopness UI Color Pallete
