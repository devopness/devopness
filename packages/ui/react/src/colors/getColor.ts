import get from 'lodash/get'

import type { Flatten, HexDigit, OpacityFromFloatToHex } from './types'

const colors = {
  amber: {
    800: '#a39015',
    500: '#fab01c',
    400: '#fdb762',
    300: '#fdd88e',
    200: '#fee2c0',
    150: '#feebc6',
    100: '#fff1e0',
    50: '#fff6ef',
    10: '#fffcf5',
  },
  black: '#000000',
  blue: {
    950: '#2e374e',
    800: '#2e364e',
    750: '#37436b',
    700: '#0064a5',
    600: '#016ee9',
    500: '#0496f5',
    400: '#96c7fe',
    300: '#b9daff',
    100: '#dcecff',
    50: '#e8f6ff',
    25: '#eef2ff',
  },
  brown: {
    400: '#b5a7a2',
    50: '#faf8f1',
  },
  cyan: {
    800: '#537e8c',
    400: '#00d7d7',
    300: '#02e0d1',
    200: '#e0f7fa',
    100: '#ebfafb',
    50: '#f6fffe',
  },
  fuchsia: {
    600: '#d000ff',
    10: '#fdf5ff',
  },
  gray: {
    900: '#42495a',
    800: '#828795',
    615: '#9198a5',
    600: '#9196a4',
    500: '#b0b3bc',
    400: '#afb8c9',
    300: '#bdc4d3',
    200: '#f3f7fe',
  },
  green: {
    800: '#57b261',
    600: '#0cd356',
    300: '#9eedbb',
    200: '#b6f2cc',
    150: '#cef6dd',
    125: '#effff1',
    100: '#edf9ee',
  },
  indigo: {
    100: '#f5f6ff',
    50: '#f6f8ff',
    10: '#f8faff',
  },
  orange: {
    700: '#f89532',
    600: '#ff8700',
    500: '#f8bf4d',
    400: '#fdba62',
    100: '#f6f4e7',
    50: '#fff6f0',
    10: '#fff9f0',
  },
  purple: {
    800: '#786efd',
    400: '#cac7fc',
    300: '#dfe4fd',
    275: '#dfe6ff',
    250: '#e6ecff',
    100: '#f2f1ff',
    50: '#f2f2fa',
  },
  red: {
    500: '#fd595c',
    450: '#ff6666',
    400: '#f67771',
    300: '#febdbe',
    200: '#fecdce',
    150: '#ffdede',
    100: '#ffe8e8',
    50: '#fff3f3',
  },
  slate: {
    600: '#4a526a',
    450: '#72798f',
    400: '#abb4c5',
    300: '#e3e8f9',
  },
  stone: {
    100: '#f6f6f6',
    50: '#edf2f3',
  },
  transparent: 'transparent',
  white: '#ffffff',
  yellow: {
    300: '#fff793',
  },
} as const

/**
 * Maps Color paths on format <color>.<shade> to its value in hex, based on colors object
 */
type ColorToHexMapper = Flatten<typeof colors>

/**
 * Lists all valid color codes, following format <color>.<shade>
 */
type Color = keyof ColorToHexMapper

/**
 * Gets the color' hex value
 *
 * @param name Color name on format <color>.<shade>
 */
const getColor = <TColor extends Color>(
  name: TColor
): ColorToHexMapper[TColor] =>
  get(colors, name) as unknown as ColorToHexMapper[TColor]

/**
 * Adds opacity suffix to colors
 *
 * @param color string in hexadecimal format #rrggbb
 *
 * @param opacity number in range 0.0 to 1.0
 *
 * This methods transform opacity from number to a hexadecimal (16) color alpha component, meaning
 * a number between 0 and ff (255) in base 16
 *
 * @returns color with added opacity information #rrggbbaa
 *
 * @example
 *
 * > getOpacity(#fff1e0, 0.5) // => #fff1e07f, 7f == 0.5
 *
 * @see {@link https://developer.mozilla.org/en-US/docs/Web/CSS/hex-color#a}
 */
const getOpacity = <
  HexColor extends ColorToHexMapper[Color],
  Opacity extends number,
>(
  color: HexColor,
  opacity: Opacity
) => {
  if (!color.startsWith('#')) return color as Exclude<HexColor, `#${string}`>

  return `${color}${(opacity * 255).toString(16).slice(0, 2)}` as `${Extract<HexColor, `#${string}`>}${Opacity extends keyof OpacityFromFloatToHex
    ? OpacityFromFloatToHex[Opacity]
    : `${HexDigit}${HexDigit}`}`
}

export type { Color }

export { getColor, getOpacity }
