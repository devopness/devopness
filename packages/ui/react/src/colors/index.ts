export * from './getColor'
export * from './types'
