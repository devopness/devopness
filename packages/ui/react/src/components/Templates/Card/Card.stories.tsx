import type { Meta, StoryObj } from '@storybook/react-vite'

import { Card } from './Card'

const meta = {
  component: Card,
  argTypes: {
    indicator: { control: 'number' },
    icon: { control: 'text' },
  },
} satisfies Meta<typeof Card>

type Story = StoryObj<typeof meta>

/**
 * Default story showing basic card usage
 */
const Default: Story = {
  args: {
    title: 'Example Card',
    subtitle: 'Card subtitle',
    avatarProps: {
      backgroundColor: 'purple.800',
    },
    headerProps: {
      backgroundColor: 'white',
    },
    url: '#',
  },
}

/**
 * Story demonstrating card with an icon instead of letter avatar
 */
const WithIcon: Story = {
  args: {
    ...Default.args,
    icon: 'github',
    title: 'GitHub Repository',
  },
}

/**
 * Story showing card with indicator
 */
const WithIndicator: Story = {
  args: {
    ...Default.args,
    indicator: 5,
    title: 'Notifications',
  },
}

/**
 * Story demonstrating card with footer actions
 */
const WithFooter: Story = {
  args: {
    ...Default.args,
    title: 'Card with Actions',
    footer: [
      {
        icon: 'edit',
        label: 'Edit',
        url: '#edit',
        tooltip: 'Edit item',
      },
      {
        icon: {
          name: 'delete',
          color: 'red.500',
        },
        label: 'Delete',
        url: '#delete',
        tooltip: 'Delete item',
      },
    ],
  },
}

export default meta
export { Default, WithIcon, WithIndicator, WithFooter }
