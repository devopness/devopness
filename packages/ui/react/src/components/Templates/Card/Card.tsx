import {
  StyledAvatar,
  StyledContainer,
  StyledFooter,
  StyledFooterLabel,
  StyledHeader,
  StyledIndicator,
  StyledLetterAvatar,
  StyledSubtitle,
  StyledTitle,
} from './Card.styled'
import type { Color } from 'src/colors'
import { Button } from 'src/components/Buttons/Button'
import type { IconProps } from 'src/components/Primitives/Icon'
import { Icon } from 'src/components/Primitives/Icon'
import type { LinkProps } from 'src/components/Primitives/Link'
import { Link } from 'src/components/Primitives/Link'
import type { TooltipProps } from 'src/components/Primitives/Tooltip'
import { Tooltip } from 'src/components/Primitives/Tooltip'

type CardProps = React.PropsWithChildren<{
  /**
   * Props for the avatar wrapper
   */
  avatarProps: {
    backgroundColor: Color
  }
  /**
   * Array of footer items to be rendered at the bottom of the card
   */
  footer?: {
    /**
     * Icon to display next to the label
     */
    icon?: IconProps['name'] | IconProps
    /**
     * Text to display in the footer item
     */
    label?:
      | string
      | {
          content: string
          style?: React.CSSProperties
        }
    /**
     * Tooltip text to show on hover
     */
    tooltip?: string | Omit<TooltipProps, 'children'>
    /**
     * URL to navigate to when footer item is clicked
     */
    url?: LinkProps['to'] | Omit<LinkProps, 'style'>
  }[]
  /**
   * Props for the header
   */
  headerProps?: {
    backgroundColor?: Color
    borderBottomColor?: Color
  }
  /**
   * Icon to display in the card
   */
  icon?: IconProps['name'] | IconProps
  /**
   * Unique identifier for the card
   */
  id?: string
  /**
   * Indicator to display in the card
   */
  indicator?: React.ReactNode
  /**
   * Subtitle to display in the card
   */
  subtitle?: string
  /**
   * Title to display in the card
   */
  title: string
  /**
   * Props for the title
   */
  titleProps?: Pick<React.HTMLProps<HTMLSpanElement>, 'style'>
  /**
   * URL to navigate to when card is clicked
   */
  url?: LinkProps['to'] | Omit<LinkProps, 'style'>
}>

/**
 * Renders an indicator for the card
 */
const Indicator = ({ indicator }: Pick<CardProps, 'indicator'>) => {
  if (indicator === undefined) return undefined

  if (typeof indicator === 'number' && indicator > 99) {
    return <StyledIndicator>+99</StyledIndicator>
  }

  return <StyledIndicator>{indicator}</StyledIndicator>
}

const DEFAULT_ICON_COLOR = 'white' satisfies Color
const DEFAULT_FOOTER_ICON_COLOR = 'gray.615' satisfies Color
const DEFAULT_ICON_SIZE = 23

/**
 * A flexible and customizable card component for displaying content
 * with a consistent, structured layout.
 *
 * @example
 * // Basic usage
 * <Card
 *   title="Environment"
 *   subtitle="Overview of current environments"
 *   avatarProps={{ backgroundColor: 'blue.500' }}
 *   icon="cubes"
 * />
 *
 * @example
 * // Card with footer and navigation
 * <Card
 *   title="Teams"
 *   url="/teams"
 *   footer={[
 *     {
 *       label: 'View All',
 *       url: '/teams'
 *     }
 *   ]}
 * />
 */
const Card = ({ children, ...props }: CardProps) => (
  <StyledContainer
    $disableMinHeight={!props.footer && !children}
    id={props.id}
  >
    <Link
      hideExternalUrlIcon
      hideUnderline
      hideUnderlineOnHover
      style={{
        display: 'block',
        marginRight: 'auto',
        width: '100%',
      }}
      {...(typeof props.url === 'object' ? props.url : undefined)}
      to={typeof props.url === 'object' ? props.url.to : props.url}
    >
      <StyledHeader
        data-testid="card-header"
        $backgroundColor={props.headerProps?.backgroundColor}
        $borderBottomColor={props.headerProps?.borderBottomColor}
        $hideBorder={!props.footer && !children}
      >
        <StyledAvatar $backgroundColor={props.avatarProps.backgroundColor}>
          {typeof props.icon === 'object' && (
            <Icon
              {...props.icon}
              color={props.icon.color ?? DEFAULT_ICON_COLOR}
              size={props.icon.size ?? DEFAULT_ICON_SIZE}
            />
          )}
          {typeof props.icon === 'string' && (
            <Icon
              name={props.icon}
              color={DEFAULT_ICON_COLOR}
              size={DEFAULT_ICON_SIZE}
            />
          )}
          {!props.icon && (
            <StyledLetterAvatar>
              {props.title.charAt(0).toUpperCase()}
            </StyledLetterAvatar>
          )}
        </StyledAvatar>
        <div>
          <Tooltip
            title={props.title}
            enableOnlyWithEllipsisPoints
          >
            <StyledTitle {...props.titleProps}>{props.title}</StyledTitle>
          </Tooltip>
          {props.subtitle && <StyledSubtitle>{props.subtitle}</StyledSubtitle>}
        </div>
        <Indicator indicator={props.indicator} />
      </StyledHeader>
    </Link>
    {children}
    {props.footer && (
      <StyledFooter
        data-testid="card-footer"
        $showBorder={!!children}
        className={props.footer.length === 1 ? 'only-child' : undefined}
      >
        {props.footer.map((footer, index) => (
          <Tooltip
            key={`footer-${String(index)}`}
            {...(typeof footer.tooltip === 'object'
              ? footer.tooltip
              : undefined)}
            title={
              typeof footer.tooltip === 'object'
                ? footer.tooltip.title
                : footer.tooltip
            }
          >
            <Link
              hideExternalUrlIcon
              hideUnderline
              style={{
                alignItems: 'center',
                display: 'flex',
                gap: '0.5rem',
                marginRight: 'auto',
              }}
              {...(typeof footer.url === 'object' ? footer.url : undefined)}
              to={typeof footer.url === 'object' ? footer.url.to : footer.url}
            >
              {typeof footer.icon === 'object' && (
                <Icon
                  {...footer.icon}
                  color={footer.icon.color ?? DEFAULT_FOOTER_ICON_COLOR}
                />
              )}
              {typeof footer.icon === 'string' && (
                <Icon
                  name={footer.icon}
                  color={DEFAULT_FOOTER_ICON_COLOR}
                />
              )}
              <Button
                buttonType="borderless"
                noMargin
                noPadding
              >
                {typeof footer.label === 'object' && (
                  <StyledFooterLabel {...footer.label}>
                    {footer.label.content}
                  </StyledFooterLabel>
                )}
                {typeof footer.label === 'string' && (
                  <StyledFooterLabel>{footer.label}</StyledFooterLabel>
                )}
              </Button>
            </Link>
          </Tooltip>
        ))}
      </StyledFooter>
    )}
  </StyledContainer>
)

export type { CardProps }
export { Card }
