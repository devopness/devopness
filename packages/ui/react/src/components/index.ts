export * from './Buttons'
export * from './Headers'
export * from './Templates'
