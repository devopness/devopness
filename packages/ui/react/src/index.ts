export * from './stories/Button'
export * from './stories/Header'
