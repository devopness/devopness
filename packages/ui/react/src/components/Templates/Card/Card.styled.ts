import { styled } from 'styled-components'

import type { CardProps } from './Card'
import { getColor } from 'src/colors'
import type { TransientProps } from 'src/components/types'
import { getFont } from 'src/fonts'

type StyledContainerProps = {
  /**
   * Disable the minimum height of the container
   *
   * @default false
   */
  $disableMinHeight?: boolean
}

const StyledContainer = styled.div<StyledContainerProps>`
  display: flex;
  min-height: ${({ $disableMinHeight }) =>
    $disableMinHeight ? 'auto' : '105px'};
  width: relative;
  border-radius: 10px;
  border: 1px solid ${getColor('slate.300')};
  background-color: ${getColor('white')};
  box-sizing: border-box;
  font-family: 'Roboto';
  overflow: hidden;
  flex-direction: column;
  align-items: center;
`

const StyledAvatar = styled.div<TransientProps<CardProps['avatarProps']>>`
  align-items: center;
  background-color: ${({ $backgroundColor }) => getColor($backgroundColor)};
  border-radius: 5px;
  box-sizing: border-box;
  display: flex;
  grid-area: icon;
  justify-content: center;
  height: 36px;
  width: 36px;
`

const StyledTitle = styled.span`
  color: ${getColor('blue.950')};
  font-size: 16px;
  font-weight: bold;
  grid-area: title;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
`

const StyledSubtitle = styled.span`
  color: ${getColor('gray.615')};
  font-size: 13px;
  grid-area: subtitle;
  max-width: calc(100% - 20px);
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
`

const StyledIndicator = styled.span`
  color: ${getColor('blue.950')};
  font-size: 16px;
  font-weight: bold;
  grid-area: indicator;
  text-align: right;
`

type StyledHeaderProps = TransientProps<CardProps['headerProps']> & {
  /**
   * Hide the border of the header
   *
   * @default false
   */
  $hideBorder?: boolean
}

const StyledHeader = styled.div<StyledHeaderProps>`
  align-items: center;
  background-color: ${({ $backgroundColor }) =>
    getColor($backgroundColor ?? 'transparent')};
  border-bottom: ${({ $borderBottomColor, $hideBorder }) =>
    $hideBorder
      ? 'none'
      : `1px solid ${getColor($borderBottomColor ?? 'slate.300')}`};
  box-sizing: border-box;
  color: ${getColor('gray.615')};
  display: grid;
  grid-gap: 0 15px;
  grid-template-areas: 'icon title indicator';
  grid-template-columns: 36px calc(100% - 85px) 20px;
  grid-template-rows: 36px;
  height: 63px;
  padding: 14px 15px 0 15px;
  width: 100%;
`

type StyledFooterProps = {
  /**
   * Show the footer border
   *
   * @default false
   */
  $showBorder?: boolean
}

const StyledFooter = styled.div<StyledFooterProps>`
  align-items: center;
  box-sizing: border-box;
  display: flex;
  flex-direction: row;
  height: 42px;
  /* If there are 2 or more direct children, use space-between */
  justify-content: space-between;
  max-width: 100%;
  padding-bottom: 15px;
  /**
   * The padding-left/right should be the same as StyledHeader's padding-left/right
   */
  padding-left: 15px;
  padding-right: 15px;
  padding-top: 15px;
  width: 100%;

  border-top: ${({ $showBorder }) =>
    $showBorder ? `1px solid ${getColor('slate.300')}` : 'none'};

  /**
   * When the footer has only one child, add a hover effect to footer
   *
   * Using classname "only-child" here is a workaround for testing environment compatibility
   * Current testing setup does not recognize ":has(:only-child)" as a valid CSS selector
   */
  &.only-child {
    padding-left: 0;

    &:hover {
      background-color: ${getColor('indigo.10')};
    }
  }

  /* When there's only one child, make it fill the entire row */
  > *:only-child {
    width: 100%;

    /**
     * When the child is a link, make it fill the entire row
     */
    & > a {
      display: flex;
      justify-content: center;
      width: 100%;
    }
  }
`

const StyledFooterLabel = styled.span`
  text-transform: uppercase;
`

const StyledLetterAvatar = styled.span`
  font-size: 23px;
  font-weight: 700;
  font-family: ${getFont('roboto')};
  color: ${getColor('blue.950')};
`

export {
  StyledAvatar,
  StyledContainer,
  StyledFooter,
  StyledFooterLabel,
  StyledHeader,
  StyledIndicator,
  StyledLetterAvatar,
  StyledSubtitle,
  StyledTitle,
}
