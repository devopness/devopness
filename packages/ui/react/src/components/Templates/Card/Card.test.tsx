import { render, screen } from '@testing-library/react'
import '@testing-library/jest-dom'
import { describe, expect, it } from 'vitest'

import { Card } from './Card'
import type { CardProps } from './Card'
import { getColor } from 'src/colors'
import { testHoverTooltip } from 'src/test-utils'

/**
 * Default props used across all Card component tests
 */
const defaultProps = {
  title: 'Test Card',
  avatarProps: {
    backgroundColor: 'purple.800',
  },
} satisfies CardProps

describe('Card', () => {
  describe('renders correctly', () => {
    it('with default props', () => {
      render(<Card {...defaultProps} />)

      expect(screen.getByText('Test Card')).toBeInTheDocument()
      expect(screen.getByText('T')).toBeInTheDocument() // First letter avatar
    })

    it('with subtitle', () => {
      render(
        <Card
          {...defaultProps}
          subtitle="Test Subtitle"
        />
      )

      expect(screen.getByText('Test Subtitle')).toBeInTheDocument()
    })

    it('without subtitle when not provided', () => {
      render(<Card {...defaultProps} />)

      expect(screen.queryByText(/subtitle/i)).not.toBeInTheDocument()
    })

    it('with custom icon instead of letter avatar', () => {
      render(
        <Card
          {...defaultProps}
          icon="add"
          url="/test"
        />
      )

      // Letter avatar should not be present when icon is provided
      expect(screen.queryByText('T')).not.toBeInTheDocument()
      // Find SVG element
      expect(document.querySelector('a[href="/test"] svg')).toBeInTheDocument()
    })

    it('with indicator', () => {
      render(
        <Card
          {...defaultProps}
          indicator={5}
        />
      )

      expect(screen.getByText('5')).toBeInTheDocument()
    })

    it('with indicator > 99 shows +99', () => {
      render(
        <Card
          {...defaultProps}
          indicator={100}
        />
      )

      expect(screen.getByText('+99')).toBeInTheDocument()
    })

    it('with custom header background color', () => {
      render(
        <Card
          {...defaultProps}
          headerProps={{
            backgroundColor: 'blue.100',
          }}
          // Without any footer, the border will be hidden
          footer={[
            {
              label: 'Settings',
              url: '/settings',
            },
          ]}
        />
      )

      const header = screen.getByTestId('card-header')
      expect(header).toBeInTheDocument()
      expect(header).toHaveStyle({
        'background-color': 'rgb(220, 236, 255)',
      })
    })

    it('with custom header border bottom color', () => {
      render(
        <Card
          {...defaultProps}
          headerProps={{
            borderBottomColor: 'blue.100',
          }}
          // Without any footer, the border will be hidden
          footer={[
            {
              label: 'Settings',
              url: '/settings',
            },
          ]}
        />
      )

      const header = screen.getByTestId('card-header')
      expect(header).toBeInTheDocument()
      expect(header).toHaveStyle({
        'border-bottom-color': 'rgb(220, 236, 255)',
      })
    })

    it('with both header background and border colors', () => {
      render(
        <Card
          {...defaultProps}
          headerProps={{
            backgroundColor: 'blue.100',
            borderBottomColor: 'purple.800',
          }}
          // Without any footer, the border will be hidden
          footer={[
            {
              label: 'Settings',
              url: '/settings',
            },
          ]}
        />
      )

      const header = screen.getByTestId('card-header')
      expect(header).toBeInTheDocument()
      expect(header).toHaveStyle({
        'background-color': 'rgb(220, 236, 255)',
        'border-bottom-color': 'rgb(120, 110, 253)',
      })
    })

    it('with children content', () => {
      render(
        <Card {...defaultProps}>
          <div data-testid="child-content">Child Content</div>
        </Card>
      )

      expect(screen.getByTestId('child-content')).toBeInTheDocument()
      expect(screen.getByText('Child Content')).toBeInTheDocument()
    })

    it('with string icon converted to object with default size', () => {
      render(
        <Card
          {...defaultProps}
          icon="add"
          url="/test"
        />
      )

      const svg = document.querySelector('a[href="/test"] svg')
      expect(svg).toBeInTheDocument()
      expect(svg).toHaveAttribute('width', '23') // Default size if not specified
    })

    it('with icon object including custom size', () => {
      render(
        <Card
          {...defaultProps}
          icon={{
            name: 'add',
            size: 24,
          }}
          url="/test"
        />
      )

      const svg = document.querySelector('a[href="/test"] svg')
      expect(svg).toBeInTheDocument()
      expect(svg).toHaveAttribute('width', '24')
    })
  })

  describe('title behavior', () => {
    it('applies custom title styles when provided', () => {
      render(
        <Card
          {...defaultProps}
          titleProps={{ style: { color: 'rgb(255, 0, 0)' } }}
        />
      )

      expect(screen.getByText('Test Card')).toHaveStyle({
        color: 'rgb(255, 0, 0)',
      })
    })
  })

  describe('accessibility', () => {
    it('applies custom id when provided', () => {
      render(
        <Card
          {...defaultProps}
          id="custom-card-id"
        />
      )

      expect(
        screen.getByText('Test Card').closest('div[id="custom-card-id"]')
      ).toBeInTheDocument()
    })

    it('makes links navigable', () => {
      render(
        <Card
          {...defaultProps}
          footer={[
            {
              label: 'view all',
              url: '/projects',
            },
          ]}
        />
      )

      const link = screen.getByText('view all').closest('a')
      expect(link).toHaveAttribute('href', '/projects')
    })
  })

  describe('footer behavior', () => {
    it('renders footer items with labels and links', () => {
      render(
        <Card
          {...defaultProps}
          footer={[
            {
              icon: 'add',
              label: 'Add',
              url: '/projects/add',
            },
            {
              label: 'view all',
              url: '/projects',
            },
          ]}
        />
      )

      const addLink = screen.getByText('Add').closest('a')
      const viewAllLink = screen.getByText('view all').closest('a')

      expect(addLink).toHaveAttribute('href', '/projects/add')
      expect(viewAllLink).toHaveAttribute('href', '/projects')
    })

    it('renders footer items with icons', () => {
      render(
        <Card
          {...defaultProps}
          footer={[
            {
              label: 'Settings',
              url: '/settings',
              icon: {
                name: 'settings',
                size: 16,
                color: 'blue.500',
              },
            },
          ]}
        />
      )

      expect(screen.getByText('Settings')).toBeInTheDocument()
      // Find an SVG element within the footer link
      const link = screen.getByText('Settings').closest('a')
      expect(link?.querySelector('svg')).toBeInTheDocument()
    })

    it('shows tooltips for footer items when provided', async () => {
      render(
        <Card
          {...defaultProps}
          footer={[
            {
              label: 'Settings',
              url: '/settings',
              tooltip: 'Manage your settings',
            },
          ]}
        />
      )

      const footerButton = screen.getByText('Settings')
      await testHoverTooltip({
        element: footerButton,
        tooltipText: 'Manage your settings',
      })
    })

    it('renders multiple footer items with mixed content', () => {
      render(
        <Card
          {...defaultProps}
          footer={[
            {
              label: 'Details',
              url: '/details',
              tooltip: 'View details',
            },
            {
              label: 'Settings',
              url: '/settings',
              icon: {
                name: 'settings',
                size: 16,
              },
            },
            {
              label: 'Delete',
              url: '/delete',
              icon: {
                name: 'delete',
                size: 16,
                color: 'red.500',
              },
              tooltip: 'Delete item',
            },
          ]}
        />
      )

      expect(screen.getByText('Details')).toBeInTheDocument()
      expect(screen.getByText('Details').closest('a')).toHaveAttribute(
        'href',
        '/details'
      )
      expect(screen.getByText('Settings')).toBeInTheDocument()
      expect(screen.getByText('Settings').closest('a')).toHaveAttribute(
        'href',
        '/settings'
      )
      expect(screen.getByText('Delete')).toBeInTheDocument()
      expect(screen.getByText('Delete').closest('a')).toHaveAttribute(
        'href',
        '/delete'
      )
    })

    it('handles object-style tooltip props in footer items', async () => {
      render(
        <Card
          {...defaultProps}
          footer={[
            {
              label: 'Settings',
              url: '/settings',
              tooltip: {
                title: 'Advanced settings',
                placement: 'top',
              },
            },
          ]}
        />
      )

      const footerButton = screen.getByText('Settings')
      await testHoverTooltip({
        element: footerButton,
        tooltipText: 'Advanced settings',
      })
    })

    it('renders footer with default icon color when not specified', () => {
      render(
        <Card
          {...defaultProps}
          footer={[
            {
              label: 'Settings',
              url: '/settings',
              icon: {
                name: 'settings',
                size: 16,
              },
            },
          ]}
        />
      )

      const link = screen.getByText('Settings').closest('a')
      const svg = link?.querySelector('svg')
      expect(svg).toHaveAttribute('color', getColor('gray.615'))
    })

    it('renders footer with string icon using default color', () => {
      render(
        <Card
          {...defaultProps}
          footer={[
            {
              label: 'Add',
              url: '/add',
              icon: 'add',
            },
          ]}
        />
      )

      const link = screen.getByText('Add').closest('a')
      const svg = link?.querySelector('svg')
      expect(svg).toHaveAttribute('color', getColor('gray.615'))
    })

    it('hides footer when no children present', () => {
      render(<Card {...defaultProps} />)

      const footer = screen.queryByTestId('card-footer')
      expect(footer).not.toBeInTheDocument()
    })
  })

  describe('border behavior', () => {
    it('shows footer border when children and footer are present', () => {
      render(
        <Card
          {...defaultProps}
          footer={[
            {
              label: 'Settings',
              url: '/settings',
            },
          ]}
        >
          <div>Child content</div>
        </Card>
      )

      const footer = screen.getByTestId('card-footer')
      expect(footer).toHaveStyle({
        borderTop: expect.any(String),
      })
    })

    it('hides footer border when no children are present', () => {
      render(
        <Card
          {...defaultProps}
          footer={[
            {
              label: 'Settings',
              url: '/settings',
            },
          ]}
        />
      )

      const footer = screen.getByTestId('card-footer')
      expect(footer).toHaveStyle({
        borderTop: '',
      })
    })
  })

  describe('url behavior', () => {
    it('makes card header clickable when url is provided', () => {
      const { rerender } = render(
        <Card
          {...defaultProps}
          url="/test-url"
        />
      )

      const link = screen.getByText('Test Card').closest('a')
      expect(link).toHaveAttribute('href', '/test-url')
      // Default target is _blank for @devopness/ui-react:Link component
      expect(link).toHaveAttribute('target', '_blank')

      rerender(
        <Card
          {...defaultProps}
          url={{
            to: '/test-url-2',
            target: '_self',
          }}
        />
      )

      expect(link).toHaveAttribute('href', '/test-url-2')
      expect(link).toHaveAttribute('target', '_self')
    })
  })
})
