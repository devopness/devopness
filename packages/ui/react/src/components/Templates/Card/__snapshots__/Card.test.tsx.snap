// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Card > renders correctly > with both header background and border colors 1`] = `
<div
  class="sc-bRKDuR eAuLul"
>
  <a
    class="sc-eqYatC jOuVVh"
    color="#786efd"
    rel="noopener noreferrer"
    style="display: block; margin-right: auto; width: 100%;"
    target="_blank"
  >
    <div
      class="sc-jwTyAe dWVCDu"
      data-testid="card-header"
    >
      <div
        class="sc-hvigdm hAPlbG"
      >
        <span
          class="sc-hwkwBN gwkCiw"
        >
          T
        </span>
      </div>
      <div>
        <span
          aria-label="Test Card"
          class="sc-kcLKEh jzBbsG translate"
          data-mui-internal-clone-element="true"
          data-testid="tooltip"
        >
          <span
            class="sc-fhHczv jwuUAs"
          >
            Test Card
          </span>
        </span>
      </div>
    </div>
  </a>
  <div
    class="sc-hjsuWn fbvJFN only-child"
    data-testid="card-footer"
  >
    <span
      aria-label=""
      class="sc-kcLKEh jzBbsG translate"
      data-mui-internal-clone-element="true"
      data-testid="tooltip"
    >
      <a
        class="sc-eqYatC bUqRXR"
        color="#786efd"
        href="/settings"
        rel="noopener noreferrer"
        style="align-items: center; display: flex; gap: 0.5rem; margin-right: auto;"
        target="_blank"
      >
        <button
          class="sc-lgpSej dlOtfd"
          data-testid="button"
        >
          <span
            class="sc-dYwGCk translate"
            data-testid="button-label"
          >
            <span
              class="sc-jJLAfE gkHyEA"
            >
              Settings
            </span>
          </span>
        </button>
      </a>
    </span>
  </div>
</div>
`;

exports[`Card > renders correctly > with custom header border bottom color 1`] = `
<div
  class="sc-bRKDuR eAuLul"
>
  <a
    class="sc-eqYatC jOuVVh"
    color="#786efd"
    rel="noopener noreferrer"
    style="display: block; margin-right: auto; width: 100%;"
    target="_blank"
  >
    <div
      class="sc-jwTyAe bAkNFh"
      data-testid="card-header"
    >
      <div
        class="sc-hvigdm hAPlbG"
      >
        <span
          class="sc-hwkwBN gwkCiw"
        >
          T
        </span>
      </div>
      <div>
        <span
          aria-label="Test Card"
          class="sc-kcLKEh jzBbsG translate"
          data-mui-internal-clone-element="true"
          data-testid="tooltip"
        >
          <span
            class="sc-fhHczv jwuUAs"
          >
            Test Card
          </span>
        </span>
      </div>
    </div>
  </a>
  <div
    class="sc-hjsuWn fbvJFN only-child"
    data-testid="card-footer"
  >
    <span
      aria-label=""
      class="sc-kcLKEh jzBbsG translate"
      data-mui-internal-clone-element="true"
      data-testid="tooltip"
    >
      <a
        class="sc-eqYatC bUqRXR"
        color="#786efd"
        href="/settings"
        rel="noopener noreferrer"
        style="align-items: center; display: flex; gap: 0.5rem; margin-right: auto;"
        target="_blank"
      >
        <button
          class="sc-lgpSej dlOtfd"
          data-testid="button"
        >
          <span
            class="sc-dYwGCk translate"
            data-testid="button-label"
          >
            <span
              class="sc-jJLAfE gkHyEA"
            >
              Settings
            </span>
          </span>
        </button>
      </a>
    </span>
  </div>
</div>
`;
